specs/hotel_spec.rb

```diff
@@ -0,0 +1,248 @@
+      @conrad.reservations.must_be_empty
+      @conrad.reservations.must_be_empty
+    end
+  end # ------------------------- describe #initialize block
+
+  describe '#list_all_rooms' do
+    it 'Returns a list of rooms' do
+      @conrad.list_all_rooms.must_be_kind_of Array
+    end
+    it 'Must contain 20 data entries' do
+      assert_equal 20, @conrad.list_all_rooms.length
+    end
+    it 'Should contain room names (String)' do
+      @conrad.list_all_rooms[0].must_be_kind_of String
+    end
+    it 'Should contain legitimate room names' do
+      name = 'Room 1'
+      @conrad.list_all_rooms.include?(name).must_equal true
+    end
+  end # ------------------------- describe #list_all_rooms block
+
+  describe '#reserve_room(checkin, checkout = nil)' do
+    it 'Adds a new Reservation' do
+      reservation = @conrad.reserve_room(@checkin, @checkout)
+      reservation.must_be_instance_of Hotels::Reservation
+      assert_equal 1, @conrad.reservations.length
+    end
+    it 'Reserves a Room for the correct number of nights' do
+      nights = @conrad.reserve_room(@checkin, @checkout).dates.length
+      assert_equal 4, nights
+    end
+    it 'Will not allow users to book a duplicate room' do
+      20.times { @conrad.reserve_room(@checkin) }
+      rooms = @conrad.reservations.map(&:rooms).flatten
+      assert_equal 20, rooms.uniq.length
+    end
+    it 'Allows users to check-in on the check-out date of 20 reservations' do
+      20.times { @conrad.reserve_room(@checkin, @checkout) }
+      @conrad.reserve_room(@checkout)
+      assert_equal 21, @conrad.reservations.length
+    end
+    it 'Will raise an error if trying to book a room on a full day' do
+      20.times { @conrad.reserve_room(@checkin, @checkout) }
+      proc {
+        @conrad.reserve_room(@checkin, @checkout)
+      }.must_raise ArgumentError
+    end
+    it 'Cannot reserve rooms that have been set for a block' do
+      4.times { @conrad.book_block(5, @checkin, @checkout) }
+      proc {
+        @conrad.reserve_room(@checkin, @checkout)
+      }.must_raise ArgumentError
+    end
+  end # ------------------------- describe #reserve_room block
+
+  describe '#reserve_by_room(checkin, checkout, room_no)' do
+    it 'Adds a new Reservation' do
+      reservation = @conrad.reserve_by_room(@checkin, @checkout, 1)
+      reservation.must_be_instance_of Hotels::Reservation
+      assert_equal 1, @conrad.reservations.length
+    end
+    it 'Reserves a Room for the correct number of nights' do
+      nights = @conrad.reserve_by_room(@checkin, @checkout, 1).dates.length
+      assert_equal 4, nights
+    end
+    it 'Will not allow users to book a duplicate Room' do
+      @conrad.reserve_by_room(@checkin, @checkout, 1)
+      proc {
+        @conrad.reserve_by_room(@checkin, @checkout, 1)
+      }.must_raise ArgumentError
+      proc {
+        @conrad.reserve_by_room(@checkin + 1, 1)
+      }.must_raise ArgumentError
+    end
```

Comment:
This is a good start, but there are a wide variety of interesting cases around whether or not a room is available that you haven't explored here. My full list includes:
- Not available:
  - Same dates
  - Overlaps in the front
  - Overlaps in the back
  - Completely contained
  - Completely containing
- Available:
  - Completely before
  - Completely after
  - Ends on the checkin date
  - Starts on the checkout date